File: daraja/src/main/java/com/androidstudy/daraja/Builder.java
package com.androidstudy.daraja;

import com.androidstudy.daraja.network.URLs;
import com.androidstudy.daraja.repo.DarajaRepository;
import com.androidstudy.daraja.util.Environment;
import com.androidstudy.daraja.constants.TransactionType;

public class Builder {
    private String consumerKey;
    private String consumerSecret;
    private String businessShortCode;
    private String passKey;
    private TransactionType transactionType;
    private String callbackUrl;
    private Environment environment;

    Builder(String consumerKey, String consumerSecret) {
        this.consumerKey = consumerKey;
        this.consumerSecret = consumerSecret;
    }

    public Builder setPassKey(String passKey) {
        this.passKey = passKey;

        return this;
    }

    public Builder setTransactionType(TransactionType transactionType) {
        this.transactionType = transactionType;

        return this;
    }

    public Builder setCallbackUrl(String callbackUrl) {
        this.callbackUrl = callbackUrl;

        return this;
    }

    public Builder setBusinessShortCode(String businessShortCode) {
        this.businessShortCode = businessShortCode;

        return this;
    }

    public Builder setEnvironment(Environment environment) {
        this.environment = environment;

        return this;
    }

    public Daraja build() {
        Daraja daraja = new Daraja();
        daraja.consumerKey = this.consumerKey;
        daraja.consumerSecret = this.consumerSecret;
        daraja.businessShortCode = this.businessShortCode;
        daraja.passKey = this.passKey;
        daraja.transactionType = this.transactionType;
        daraja.callbackUrl = this.callbackUrl;
        daraja.baseUrl = (this.environment == Environment.SANDBOX) ? URLs.SANDBOX_BASE_URL : URLs.PRODUCTION_BASE_URL;

        daraja.repo = new DarajaRepository(daraja.consumerKey, daraja.consumerSecret, daraja.baseUrl);

        return daraja;
    }
}

File: daraja/src/main/java/com/androidstudy/daraja/Builder.kt
/*
 * Copyright 2021 Juma Allan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.androidstudy.daraja

import com.androidstudy.daraja.util.TransactionType
import com.androidstudy.daraja.data.repo.DarajaRepository
import com.androidstudy.daraja.util.Environment

/**
 * Keys Builder. Creates a [Daraja] payment load.
 * @param [consumerKey] for your application
 * @param [consumerSecret] for your application
 */

class Builder(private var consumerKey: String, private var consumerSecret: String) {

    private lateinit var businessShortCode: String
    private lateinit var passKey: String
    private lateinit var transactionType: TransactionType
    private lateinit var callbackUrl: String
    private lateinit var environment: Environment

    fun setPassKey(passKey: String): Builder {
        this.passKey = passKey
        return this
    }

    fun setTransactionType(transactionType: TransactionType): Builder {
        this.transactionType = transactionType
        return this
    }

    fun setCallbackUrl(callbackUrl: String): Builder {
        this.callbackUrl = callbackUrl
        return this
    }

    fun setBusinessShortCode(businessShortCode: String): Builder {
        this.businessShortCode = businessShortCode
        return this
    }

    fun setEnvironment(environment: Environment): Builder {
        this.environment = environment
        return this
    }

    fun build(): Daraja {
        val daraja = Daraja
        daraja.consumerKey = consumerKey
        daraja.consumerSecret = consumerSecret
        daraja.businessShortCode = businessShortCode
        daraja.passKey = passKey
        daraja.transactionType = transactionType
        daraja.callbackUrl = callbackUrl
        daraja.baseUrl = environment.url

        daraja.repo = DarajaRepository(daraja.consumerKey, daraja.consumerSecret, daraja.baseUrl)

        return daraja
    }
}